relay/nakama/custom_auth.go

```diff
@@ -0,0 +1,98 @@
+package main
+
+import (
+	"context"
+	"database/sql"
+	"encoding/json"
+	"errors"
+
+	"github.com/heroiclabs/nakama-common/api"
+	"github.com/heroiclabs/nakama-common/runtime"
+	"github.com/rotisserie/eris"
+	"google.golang.org/grpc/codes"
+
+	"pkg.world.dev/world-engine/relay/nakama/siwe"
+	"pkg.world.dev/world-engine/relay/nakama/utils"
+)
+
+const (
+	signInWithEthereumType = "siwe"
+)
+
+var (
+	ErrBadCustomAuthType = errors.New("bad custom auth type")
+)
+
+func initCustomAuthentication(initializer runtime.Initializer) error {
+	if err := initializer.RegisterBeforeAuthenticateCustom(handleCustomAuthentication); err != nil {
+		return eris.Wrap(err, "failed to init siwe authentication")
+	}
+	return nil
+}
+
+func handleCustomAuthentication(
+	ctx context.Context,
+	logger runtime.Logger,
+	_ *sql.DB,
+	nk runtime.NakamaModule,
+	in *api.AuthenticateCustomRequest) (*api.AuthenticateCustomRequest, error) {
+	authType := in.GetAccount().GetVars()["type"]
+	// In the future, other authentication methods can be added here (e.g. Twitter)
+	if authType == signInWithEthereumType {
+		return handleSIWE(ctx, logger, nk, in)
+	}
+	return nil, ErrBadCustomAuthType
+}
+
+func handleSIWE(
+	ctx context.Context,
+	logger runtime.Logger,
+	nk runtime.NakamaModule,
+	in *api.AuthenticateCustomRequest,
+) (
+	*api.AuthenticateCustomRequest, error,
+) {
+	signerAddress := in.GetAccount().GetId()
+	signature := in.GetAccount().GetVars()["signature"]
+	message := in.GetAccount().GetVars()["message"]
+
+	if signerAddress == "" {
```

Comment:
should probably just check the fields individually then errors.Join(...) it

Reply:
done